### PR TITLE
feat: Merge parentBuildId metas

### PR DESCRIPTION
**Pull request by tkyi**
## Context
When a join occurs, we're currently just using the meta from the last build that triggered the join job to start. For example, if C requires A and B, if A finishes running first, B will trigger C to start after it is done. C would then be run with whatever metadata B had.
```
A 
     -->  C
B
```
We would like to merge all the metadata from all parent builds so that C can run using metadata from both A and B.

## Objective
This PR checks for multiple parent build IDs and merges the metadata for all of them.

## Related Links
Related to https://github.com/screwdriver-cd/screwdriver/issues/904
Blocked by https://github.com/screwdriver-cd/data-schema/pull/210

### CODE REVIEW

File: launch_test.go [hunk 1]
```diff
@@ -958,6 +986,74 @@ func TestFetchParentBuildMetaWriteError(t *testing.T) {
...
+	}
+	if actual["batman"] != "robin" {
+		t.Errorf("Error is wrong, got '%v', expected '%v'", actual["batman"], "robin")
+	}
+	if actual["wonder"] != "woman" {
+		t.Errorf("Error is wrong, got '%v', expected '%v'", actual["wonder"], "woman")
+	}
+}
+
+func TestMergedMetaFromParentBuildsIDs(t *testing.T) {
```
Comment (minz1027) [inline launch_test.go]: what is this test case trying to test? should be covered by `TestFetchParentBuildsMeta ` already, right?

Comment (tkyi) [reply launch_test.go]: Removed

File: launch.go [hunk 2]
```diff
@@ -213,20 +232,43 @@ func launch(api screwdriver.API, buildID int, rootDir, emitterPath, metaSpace, s
...
 	// If no parent build ID, and no parent event, skip fetch meta
-	if build.ParentBuildID == 0 && event.ParentEventID == 0 {
+	if len(parentBuildIDs) == 0 && event.ParentEventID == 0 {
 		log.Printf("This build has no Parent Build and no Parent Event, so fetching Meta is skipped")
 	} else {
-		// If has parent build, fetch from parent
-		if build.ParentBuildID != 0 {
-			log.Printf("Fetching Parent Build %d", build.ParentBuildID)
-			parentBuild, err := api.BuildFromID(build.ParentBuildID)
+		if len(parentBuildIDs) > 1 { // If has multiple parent build IDs, merge their metadata
```
Comment (FenrirUnbound) [inline launch.go]: We could also make this an `else if`

File: launch.go [hunk 3]
```diff
@@ -213,20 +232,43 @@ func launch(api screwdriver.API, buildID int, rootDir, emitterPath, metaSpace, s
 	}
 
 	metaByte := []byte("")
-	parentID := 0
 	metaFile := "meta.json" // Write to "meta.json" file
-	metaSource := ""
+	metaLog := ""
+
+	parentBuildIDs := convertToArray(build.ParentBuildID)
```
Comment (d2lam) [inline launch.go]: just a concern here: if there is no `build.ParentBuildID` then it **should** return `[]`?

File: launch_test.go [hunk 4]
```diff
@@ -33,6 +33,8 @@ const (
 	TestSHA    = "abc123"
 )
 
+var TestParentBuildIDFloat interface{} = float64(1111)
```
Comment (FenrirUnbound) [inline launch_test.go]: If it's not going to be updated, use a `const` instead of `var`

File: launch_test.go [hunk 5]
```diff
@@ -865,6 +867,32 @@ func TestFetchParentBuildMeta(t *testing.T) {
 	}
 }
 
+func TestFetchParentBuildsMetaParseError(t *testing.T) {
+	TestParentBuildIDs := []float64{1111, 2222}
+	IDs := make([]interface{}, len(TestParentBuildIDs))
```
Comment (FenrirUnbound) [inline launch_test.go]: Why can't we just initialize `IDs` with the test values?

File: launch.go (hunk 3 above)
Comment (tkyi) [reply launch.go]: Yes, I tried running it in the Golang playground and it looks like `convertToArray` should handle `nil` properly

File: launch.go (hunk 2 above)
Comment (tkyi) [reply launch.go]: We would have to duplicate the `create meta space section` or make a function out of it

File: launch.go (hunk 3 above)
Comment (FenrirUnbound) [reply launch.go]: I guess that depends on how Golang would unmarshal it if it's `0`

Comment (FenrirUnbound) [reply launch.go]: > Yes, I tried running it in the Golang playground and it looks like convertToArray should handle nil properly

What if it's `undefined`?

File: launch_test.go (hunk 5 above)
Comment (tkyi) [reply launch_test.go]: I can't seem to initialize `IDs` in a simpler way. I think interface arrays are treated differently than int arrays.

File: launch_test.go (hunk 4 above)
Comment (tkyi) [reply launch_test.go]: I can't seem to assign that var as a `const`. I'm getting:
```
./launch_test.go:38: const initializer []float64 literal is not a constant
```

Comment (FenrirUnbound) [reply launch_test.go]: Yeah, I guess it's because slices aren't immutable

Comment (FenrirUnbound) [reply launch_test.go]: It's fine as is 😺

File: launch_test.go (hunk 5 above)
Comment (FenrirUnbound) [reply launch_test.go]: Fine as-is

Comment (tkyi) [reply launch_test.go]: Moved the vars to a more global level

**Comment by tkyi** [comment]
Darren's PR broke my tests :[

**Pull request merged**